include/exadg/incompressible_navier_stokes/spatial_discretization/generalized_newtonian_model.cpp

```diff
@@ -0,0 +1,334 @@
+
+  scalar shear_rate =
+    std::sqrt(2.0 * scalar_product(symmetric_velocity_gradient, symmetric_velocity_gradient));
+
+  scalar viscosity_factor;
+  compute_viscosity_factor(viscosity_factor, shear_rate);
+
+  viscosity += viscosity_factor * viscosity_margin;
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::compute_viscosity_factor(scalar &       viscosity_factor,
+                                                                 scalar const & shear_rate) const
+{
+  switch(generalized_newtonian_model)
+  {
+    case GeneralizedNewtonianViscosityModel::Undefined:
+      AssertThrow(generalized_newtonian_model != GeneralizedNewtonianViscosityModel::Undefined,
+                  dealii::ExcMessage("parameter must be defined"));
+      break;
+    case GeneralizedNewtonianViscosityModel::GeneralizedCarreauYasuda:
+      generalized_carreau_yasuda_model(viscosity_factor, shear_rate);
+      break;
+    case GeneralizedNewtonianViscosityModel::Carreau:
+      carreau_model(viscosity_factor, shear_rate);
+      break;
+    case GeneralizedNewtonianViscosityModel::Cross:
+      cross_model(viscosity_factor, shear_rate);
+      break;
+    case GeneralizedNewtonianViscosityModel::SimplifiedCross:
+      simplified_cross_model(viscosity_factor, shear_rate);
+      break;
+    case GeneralizedNewtonianViscosityModel::PowerLaw:
+      power_law_model(viscosity_factor, shear_rate);
+      break;
+    default:
+      AssertThrow(
+        false, dealii::ExcMessage("This GeneralizedNewtonianViscosityModel is not implemented."));
+  }
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::generalized_carreau_yasuda_model(
+  scalar &       viscosity_factor,
+  scalar const & shear_rate) const
+{
+  // eta = eta_oo + (eta_0 - eta_oo) * [k + (l * y)^a]^[(n-1)/a]
+  viscosity_factor = shear_rate * lambda;
+  viscosity_factor = pow(viscosity_factor, a);
+  viscosity_factor += kappa;
+  viscosity_factor = pow(viscosity_factor, static_cast<Number>((n - 1.0) / a));
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::carreau_model(scalar &       viscosity_factor,
+                                                      scalar const & shear_rate) const
+{
+  // eta = eta_oo + (eta_0 - eta_oo) * [k + (l * y)^a]^[(n-1)/a]
+  // with k = 1 and a = 2
+  // eta = eta_oo + (eta_0 - eta_oo) * [1 + l^2 * y^2]^[(n-1)/2]
+  viscosity_factor = shear_rate * shear_rate * (lambda * lambda);
+  viscosity_factor += 1.0;
+  viscosity_factor = pow(viscosity_factor, static_cast<Number>((n - 1.0) * 0.5));
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::cross_model(scalar &       viscosity_factor,
+                                                    scalar const & shear_rate) const
+{
+  // eta = eta_oo + (eta_0 - eta_oo) * [k + (l * y)^a]^[(n-1)/a]
+  // with k = 1 and n = 1 - a
```

Comment:
I make this comment exemplarily here: Is it worth implementation the same type of function multiple times with different parameters? I would implement the function only once and pass the relevant parameters (k, n, a) to this generalized function.

Hence, I assume the reason for the present implementation must be computational efficiency since the present function does not need a second call of `std::pow()`? Could you comment on computational efficiency or the impact of this part of the code on overall efficiency?

In case it should not increase costs in a noticable way, we should realize all the different models as a single function.